mypy/checkmember.py

```diff
@@ -371,8 +371,27 @@ def validate_super_call(node: FuncBase, mx: MemberContext) -> None:
 def analyze_type_callable_member_access(name: str, typ: FunctionLike, mx: MemberContext) -> Type:
     # Class attribute.
     # TODO super?
-    ret_type = typ.items[0].ret_type
+    item = typ.items[0]
+    ret_type = item.ret_type
     assert isinstance(ret_type, ProperType)
+
+    # The following is a hack. mypy often represents types as CallableType, where the signature of
+    # CallableType is determined by __new__ or __init__ of the type (this logic is in
+    # type_object_type). Then if we ever need the TypeInfo or an instance of the type, we fish
+    # around for the return type, e.g. in CallableType.type_object. Unfortunately, this is incorrect
+    # if __new__ returns an unrelated type, but we can kind of salvage things here using
+    # CallableType.bound_args
+    self_type: Type = typ
+    if len(item.bound_args) == 1 and item.bound_args[0]:
+        bound_arg = get_proper_type(item.bound_args[0])
+        if isinstance(bound_arg, Instance) and isinstance(ret_type, Instance):
+            # Unfortunately, generic arguments have already been determined for us. We need these,
+            # see e.g. testGenericClassMethodUnboundOnClass. So just copy them over to our type.
+            # This does the wrong thing with custom __new__, see testNewReturnType15, but is
+            # no worse than previous behaviour.
+            ret_type = bound_arg.copy_modified(args=ret_type.args)
```

Comment:
Maybe you can
```suggestion
            ret_type = bound_arg.copy_modified()
```
if you change checkexpr.py's `apply_generic_arguments` expand `bound_args`, though I'm not sure if it's correct to do:
```diff
 return callable.copy_modified(
     ret_type=expand_type(callable.ret_type, id_to_type),
+    bound_args=[
+        expand_type(ba, id_to_type) if ba is not None else None
+        for ba in callable.bound_args
+    ],
     variables=remaining_tvars,
     type_guard=type_guard,
 )
```